API design: Automatically install software and scope software with labels (#20872)
fleetdm/fleet: docs/REST API/rest-api.md

> ### Delete label

Reviewer: ### Dev note

Error message returned in the API when the user attempts to delete a label that's being used as a custom target for software. 

Here's what this error message will look like in the UI: https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5493-9746&t=piN5Gu8vzYRvESPT-1

Reviewer: TODO: Remove this comment before merge

Reviewer: @RachelElysia here's the error message we chatted about on our call today.

FYI @lucasmrod

<!-- Return the following error if the user attempts to delete a label that's being used as a custom target for software: Software uses this label as a custom target. Please delete the software and try again. -->

> 

Reviewer 2: @noahtalerman just to double check, the[ UI spec has "Couldn't delete."](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5493-9746&t=dyxeJtJBwn2T6DYX-1), so that should be added by the UI, and won't be part of the API error message? Or do we want them to match.

Reviewer: Hmm, I think it makes sense to the API error message to match.

I updated the error message in the PR.

FYI @lucasmrod

### Add software

> 

Reviewer: ### Dev note

Error message returned in the API when the user attempts to add a package and Fleet can't extract the version.

Here's what this error message will look like in the UI: https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5568-4943&t=ZBrK0Ypx0LwOLF5n-1

| pre_install_query  | string | form | Query that is pre-install condition. If the query doesn't return any result, Fleet won't proceed to install. |
| post_install_script | string | form | The contents of the script to run after install. If the specified script fails (exit code non-zero) software install will be marked as failed and rolled back. |
| install | string | form | Install method. Options are `manual` (use Install software to install on each host manually) and `automatic` installed to each host on the team. Default is `manual` |
| self_service | boolean | form | Self-service software is optional and can be installed by the end user. |
| labels_include_any | array | form | _Available in Fleet Premium._ Software will only be installed/available for install on hosts that have any of these labels. |
| labels_exclude_any | array | form | _Available in Fleet Premium._ Software will only be installed/available for install on hosts that don't have any of these labels. |

> 

Reviewer 2: @noahtalerman to add a software package for ["All hosts"](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5417-4753&t=dyxeJtJBwn2T6DYX-1), the UI should just omit both of these fields?

> 

Reviewer: Hey @jacobshandling! 

Currently, one can't add a software package for all hosts (all teams).

The UI shows this in the Add software modal (from dogfood):

![Screenshot 2024-08-08 at 2 12 58 PM](https://github.com/user-attachments/assets/ba8f332b-42ba-4df7-b407-74b3b246b520)

Just double checking, the `POST / software/package` API returns an error today right? (when you don't specify a team)

> 

Reviewer 2: @noahtalerman what I'm asking about is where FIgma specifies an `[All hosts` Target option](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5417-4753&t=dyxeJtJBwn2T6DYX-1). Seems like that terminology is confusing so maybe worth updating...should this mean "All team hosts"?

So the question is, what's the API expectation for "no label restrictions". Excluding both `labels_include_any` and `labels_exclude_any` seems intuitive.

> 

Reviewer: > what I'm asking about is where Figma specifies an [`All hosts` Target option](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5417-4753&t=dyxeJtJBwn2T6DYX-1).

@jacobshandling I see what you're saying. I think let's keep the copy as wireframed.

We use the same "All hosts" language in the "Add profile" modal:
![Screenshot 2024-08-09 at 2 53 26 PM](https://github.com/user-attachments/assets/9de99402-56e4-4349-8013-321e70166c69)

> what's the API expectation for "no label restrictions". Excluding both `labels_include_any` and `labels_exclude_any` seems intuitive.

Do you know what do for `POST /configuration_profiles` today? I think let's be consistent in the API.

> 

Reviewer 2: There we just omit both arrays for all hosts, so I'll do that here. Thanks!

> 

Reviewer: Sweet!

FYI @lucasmrod 

(hiding these comments now that the convo is resolved)

| install_script  | string | form | Command that Fleet runs to install software. If not specified Fleet runs [default install command](https://github.com/fleetdm/fleet/tree/f71a1f183cc6736205510580c8366153ea083a8d/pkg/file/scripts) for each package type. |
| pre_install_query  | string | form | Query that is pre-install condition. If the query doesn't return any result, Fleet won't proceed to install. |
| post_install_script | string | form | The contents of the script to run after install. If the specified script fails (exit code non-zero) software install will be marked as failed and rolled back. |

> | install | string | form | Install method. Options are `manual` (use Install software to install on each host manually) and `automatic` installed to each host on the team. Default is `manual` |

Reviewer 2: Discussing with @RachelElysia, we both think `install_type` would be clearer, does that work @noahtalerman? cc @lucasmrod

Reviewer: Hey @RachelElysia and @jacobshandling I see what y'all are saying here.

I'm on board w/ `install_type`. I updated the PR.

`*_type` seems to be consistent w/ other Fleet API endpoints. For example `label_membership_type` in "Add label": https://fleetdm.com/docs/rest-api/rest-api#add-label

@spokanemac what do they call this in Munki? Looks like they call it `INSTALL` in installomator: https://github.com/Installomator/Installomator/wiki/Configuration-and-Variables#install-behavior-force-installation

Reviewer: Marking this conversation as resolved to clean up the PR.

Content-Disposition: form-data; name="team_id"
1
--------------------------d8c247122f594ba0
Content-Disposition: form-data; name="install"
automatic
--------------------------d8c247122f594ba0

> Content-Disposition: form-data; name="labels_exclude_all"

Reviewer 3: labels_exclude_any

Content-Disposition: form-data; name="team_id"
1
--------------------------d8c247122f594ba0

> Content-Disposition: form-data; name="install"

Author: ```suggestion
Content-Disposition: form-data; name="install_type"
```

automatic
--------------------------d8c247122f594ba0

> Content-Disposition: form-data; name="labels_exclude_all"

Author: ```suggestion
Content-Disposition: form-data; name="labels_exclude_any"
```

   "host_display_name": "Marko's MacBook Pro",
   "status": "failed",

>    "output": "Installing software...\nError: The operation can’t be completed because the item “Falcon” is in use.",

Reviewer: ### Dev note

For automatic apps, if Fleet can't extract a version from the app to double check if an older version is installed, then Fleet will mark the install as failed with this copy: Failed to verify: The install script finished with exit code 0 but Fleet couldn't extract the version from FalconSensor-6.44.pkg.

For manual apps, if Fleet can't extract a version from the app to double check if the exact version is installed, then Fleet will mark the install as failed with the same copy.

This is also shown in the UI [here](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5503-10100&t=piN5Gu8vzYRvESPT-1).

Reviewer: ### Dev note

For automatic apps, if Fleet finds that the software is missing or an older version is installed after reinstalling (retrying) once, then Fleet will mark the software as failed with this copy: 

"Failed to verify: The install script finished with exit code 0 but the software is now missing. Fleet already tried to reinstall once."

This is shown in the UI [here](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5503-10126&t=piN5Gu8vzYRvESPT-1).

Reviewer: Hey @lucasmrod and @RachelElysia heads up, here's the new `"detail"` field to expose the new types of "Failed" to the IT admin.

Note that we also have a similar `"detail"` field in the `profiles` list in the `GET /hosts/:id` endpoint.

cc @gillespi314

Reviewer 4: ^ @jacobshandling updates for your work on the host details page

Reviewer: Hey @lucasmrod @jacobshandling heads up that @roperzh doesn't have concerns w/ erroring if we can't extract the version.

The plan is to move forward with adding this erorr:
![Screenshot 2024-08-09 at 4 26 52 PM](https://github.com/user-attachments/assets/fc3a8e23-f8da-4e40-84a7-3bb16a23ba02)

Reviewer 3: @noahtalerman Do we want this whole error string (in red) to be on the API? (this change is missing from this PR).

Reviewer 2: Was just looking into this as well @lucasmrod  – [had a similar conversation around deleting labels,](https://github.com/fleetdm/fleet/pull/20872#discussion_r1712369117) my guess is we want to include the whole string based on that.

Reviewer 2: Oh, but the link component will have to be added by the UI – would the API message then be `"Couldn't add. Fleet couldn't read the version from <sw_package_name>."`?

Reviewer 3: No worries. Backend has access to the filename ("<sw_package_name>").

Reviewer 3: Yeah. Last part ("Learn more") should be added by UI AFAICS.

Reviewer: > Do we want this whole error string (in red) to be on the API?

@lucasmrod yep. Thanks for catching that it's not in the PR. I added it here: https://github.com/fleetdm/fleet/pull/20872/files#r1714469004

   "host_display_name": "Marko's MacBook Pro",
   "status": "failed",
   "output": "Installing software...\nError: The operation can’t be completed because the item “Falcon” is in use.",

>    "detail": "The install script finished with exit code 0 but Fleet couldn't extract the version from the FalconSensor-6.44.pkg package.",

Reviewer 3: With the "version restriction" now added when uploading installers this error code ("Fleet couldn't extract the version from the FalconSensor-6.44.pkg package") won't happen.

Reviewer: Ah, good catch. I updated this error message to the error message when Fleet has retried once already:

Failed to verify: The install script finished with exit code 0 but the software is now missing. Fleet already tried to reinstall once.

>    "detail": "The install script finished with exit code 0 but Fleet couldn't extract the version from the FalconSensor-6.44.pkg package.",

Reviewer: ```suggestion
   "detail": "Failed to verify: The install script finished with exit code 0 but the software is now missing. Fleet already tried to reinstall once.",
```

      "installer_id": 23,
      "team_id": 3,
      "uploaded_at": "2024-04-01T14:22:58Z",
      "install": "automatic"
      "install_script": "sudo installer -pkg /temp/FalconSensor-6.44.pkg -target /",
      "pre_install_query": "SELECT 1 FROM macos_profiles WHERE uuid='c9f4f0d5-8426-4eb8-b61b-27c543c9d3db';",
      "post_install_script": "sudo /Applications/Falcon.app/Contents/Resources/falconctl license 0123456789ABCDEFGHIJKLMNOPQRSTUV-WX",
      "self_service": true,
      "labels_exclude_any": [

>         "Label name 1",

Reviewer 4: This is the way profiles works for GET, it includes an ID. Is that what the GET request will look like here too?

<img width="1218" alt="Screenshot 2024-08-07 at 11 24 46 AM" src="https://github.com/user-attachments/assets/c222281f-95b4-4c76-a67f-a06939658bde">

>         "Label name 1",

Reviewer: Hey @RachelElysia we want to return the label `id` in this endpoint too. I just updated the PR.

That said, I think the frontend (Figma [here](https://www.figma.com/design/4pfUOYy7IyMIrjMH2fuCdU/%2319551-Automatically-install-software-and-scope-software-with-labels?node-id=5417-5368&t=3aZ5dEQuNIK0oT2V-1)) will still use the label `name`:

![Screenshot 2024-08-07 at 2 35 53 PM](https://github.com/user-attachments/assets/4296484a-5583-413d-9262-f3e927c9ddaf)

I think the backend will be responsible for making sure the label `name` in this API endpoint gets changed when the IT admin changes the label's name. 

@lucasmrod does that sound right? 

Heads up that we've made a mistake when building similar features. We shipped a bug in which we aren't updating the label name for profiles: 
- #21163

>         "Label name 1",

Reviewer: @lucasmrod just giving you an extra ping here.

> the backend will be responsible for making sure the label name in this API endpoint gets changed when the IT admin changes the label's name.

When you get the chance, can you please check to see if this works? Thanks :)

>         "Label name 1",

Reviewer 3: Yes, unlike the implementation for configuration profiles we will store the `label_id` instead of `label_name` on the DB table.

>         "Label name 1",

Reviewer: Thanks Lucas! Hiding these comments are resolved to clean up the PR.

        "Label name 2"
      ],
      "status": {
        "verified": 5,
        "verifying": 3,
        "pending": 1,

>         "blocked": 1,

Reviewer 4: For the GET, are we returning an object similar to profiles?

e.g.
```
      "labels_include_all": [
        {
          "name": "macOS Version 14.x.x",
          "broken": true
        }
      ]
      ```

>         "blocked": 1,

Reviewer 4: @noahtalerman @marko-lisica 

Are we doing this for just software package and not app store app? e.g. Are we sending back verified, verifying, blocked for VPP? Are we keeping the old UI for VPP apps?

>         "blocked": 1,

Reviewer 4: Answer from @noahtalerman We are sending back verified, verifying, blocked for VPP as well, will need updates as well

>         "blocked": 1,

Reviewer: > We are sending back verified, verifying, blocked for VPP as well

Hey @RachelElysia I just called this out in the dev note here: https://github.com/fleetdm/fleet/pull/20872/files#r1697994003

Is the dev note clear enough?

cc @lucasmrod

>         "blocked": 1,

Reviewer 4: @noahtalerman it's super clear now, thank you!!

>         "blocked": 1,

Reviewer: Sweet! Hiding our comments as resolved so the PR is easier to read. Dev notes aren't lose in older comments.

      "installer_id": 23,
      "team_id": 3,
      "uploaded_at": "2024-04-01T14:22:58Z",
      "install": "automatic"
      "install_script": "sudo installer -pkg /temp/FalconSensor-6.44.pkg -target /",
      "pre_install_query": "SELECT 1 FROM macos_profiles WHERE uuid='c9f4f0d5-8426-4eb8-b61b-27c543c9d3db';",
      "post_install_script": "sudo /Applications/Falcon.app/Contents/Resources/falconctl license 0123456789ABCDEFGHIJKLMNOPQRSTUV-WX",
      "self_service": true,
      "labels_exclude_any": [
        {
          "name": "Label name 1",
          "id": 1
        },
        {
          "name": "Label name 2",
          "id": 2
        }
      ],

>       "status": {

Reviewer: ### Dev note

We can break this API endpoint by removing the `"installed"` status because app management features are tagged as experimental: https://github.com/fleetdm/fleet/pull/21164/files

More on experimental features in the handbook here: https://fleetdm.com/handbook/company/product-groups#experimental-features

>       "status": {

Reviewer: ### Dev note

When a any host enrolls to Fleet (macOS, Windows, Linux, iOS/iPadOS), we want to clear it's software `status` and cancel any pending software installs.

Also, when a host is deleted in Fleet (via the UI/API), we want to clear it's software `status` and cancel any pending software installs.

Reviewer: @lucasmrod heads up, I added what we discussed during design review plus the "when a host is deleted"

Please let me know if you have any concerns with this.

cc @gillespi314

Reviewer: FYI I closed the following bug since we're addressing it as part of this story: https://github.com/fleetdm/fleet/issues/21075

      "installer_id": 23,
      "team_id": 3,
      "uploaded_at": "2024-04-01T14:22:58Z",
      "install": "automatic"
      "install_script": "sudo installer -pkg /temp/FalconSensor-6.44.pkg -target /",
      "pre_install_query": "SELECT 1 FROM macos_profiles WHERE uuid='c9f4f0d5-8426-4eb8-b61b-27c543c9d3db';",
      "post_install_script": "sudo /Applications/Falcon.app/Contents/Resources/falconctl license 0123456789ABCDEFGHIJKLMNOPQRSTUV-WX",
      "self_service": true,
      "labels_exclude_any": [

>         {

Reviewer 2: FYI @fleetdm/frontend, I'll add the following interface for this object, which the other Label interfaces will extend:
```
export interface ILabelIdentifier {
  id: number;
  name: string;
}
```
